PWGHF/HFC/TableProducer/correlatorDMesonPairsTesting.cxx

```diff
@@ -0,0 +1,783 @@
+    for (const auto& particle : mcParticles) {
+      // check if the particle is D0 or D0bar
+      if (std::abs(particle.pdgCode()) != Pdg::kD0) {
+        continue;
+      }
+      if (abs(particle.y()) > yCandMax) {
+        continue;
+      }
+      if (ptCandMin >= 0. && particle.pt() < ptCandMin) {
+        continue;
+      }
+      auto particleType = assignParticleTypeD0Gen(particle); // Candidate type attribution
+      bool isDParticle = isTrueD(particleType);
+      bool isDbarParticle = isTrueDbar(particleType);
+      if (isDParticle) {
+        nDevent++;
+      }
+      if (isDbarParticle) {
+        nDbarevent++;
+      }
+      if (isDParticle && isDbarParticle) {
+        nDDbarevent++;
+      }
+      if (isDParticle || isDbarParticle) {
+        nDorDbarevent++;
+      }
+    }
+    if (nDevent > 0) {
+      registry.fill(HIST("hInputCheckD0McGen"), nDevent);
+    }
+    if (nDbarevent > 0) {
+      registry.fill(HIST("hInputCheckD0barMcGen"), nDbarevent);
+    }
+    if (nDDbarevent > 0) {
+      registry.fill(HIST("hInputCheckD0AndD0barMcGen"), nDDbarevent);
+    }
+    if (nDorDbarevent > 0) {
+      registry.fill(HIST("hInputCheckD0OrD0barMcGen"), nDorDbarevent);
+    }
+
+    auto massD = MassD0;
+    auto massDbar = MassD0Bar;
+
+    for (const auto& particle1 : mcParticles) {
+      // check if the particle is D0 or D0bar
+      if (std::abs(particle1.pdgCode()) != Pdg::kD0) {
+        continue;
+      }
+      registry.fill(HIST("hPtCandMcGen"), particle1.pt());
+
+      if (abs(particle1.y()) > yCandMax) {
+        continue;
+      }
+      if (ptCandMin >= 0. && particle1.pt() < ptCandMin) {
+        continue;
+      }
+      registry.fill(HIST("hEtaMcGen"), particle1.eta());
+      registry.fill(HIST("hPhiMcGen"), particle1.phi());
+      registry.fill(HIST("hPtCandAfterCutMcGen"), particle1.pt());
+
+      auto particleType1 = assignParticleTypeD0Gen(particle1); // Candidate sign attribution
+      bool isDParticle1 = isTrueD(particleType1);
+      bool isDbarParticle1 = isTrueDbar(particleType1);
+
+      // check if it's MC matched
+      int8_t matchedGen1 = particle1.flagMcMatchGen();
+      // check origin
+      int8_t originGen1 = particle1.originMcGen();
+
+      // Fill selection status single particle
+      registry.fill(HIST("hStatusSinglePartMcGen"), 1);
+      if (isDParticle1 && !isDbarParticle1) {
+        registry.fill(HIST("hStatusSinglePartMcGen"), 2);
+      } else if (isDbarParticle1 && !isDParticle1) {
+        registry.fill(HIST("hStatusSinglePartMcGen"), 3);
+      } else if (isDParticle1 && isDbarParticle1) {
+        registry.fill(HIST("hStatusSinglePartMcGen"), 4);
+      }
+
+      for (const auto& particle2 : mcParticles) {
+        // check if the particle is D0 or D0bar
+        if (std::abs(particle2.pdgCode()) != Pdg::kD0) {
+          continue;
+        }
+        // avoid double counting
+        if (particle1.mRowIndex >= particle2.mRowIndex) {
```

Comment:
In principle it would be better to do it using the iterators in the loops (it speeds up quite substantially the execution, since the unwanted combinations are not even considered in the loop). For instance, you can see in the `trackIndexSkimCreator` code: https://github.com/AliceO2Group/O2Physics/blob/82ed3f06c22f161e032fbc356f0ef07ac6ae397d/PWGHF/TableProducer/trackIndexSkimCreator.cxx#L2346